SimPEG/electromagnetics/utils/static_instrument/base.py

```diff
@@ -0,0 +1,301 @@
+
+    def make_thicknesses(self):
+        if "dep_top" in self.xyz.layer_params:
+            return np.diff(self.xyz.layer_params["dep_top"].values)
+        return get_vertical_discretization_time(
+            self.times,
+            sigma_background=0.1,
+            n_layer=self.n_layer-1
+        )
+        
+    def make_survey(self):
+        times = self.times
+        systems = [
+            self.make_system(
+                idx,
+                self.xyz.flightlines.loc[
+                    idx, [self.xyz.x_column, self.xyz.y_column, self.xyz.alt_column]
+                ].astype(float).values,
+                times)
+            for idx in range(0, len(self.xyz.flightlines))]
+        return tdem.Survey([
+            source
+            for sources in systems
+            for source in sources])
+
+    def n_param(self, thicknesses):
+        return (len(thicknesses)+1)*len(self.xyz.flightlines)
+    
+    def make_simulation(self, survey, thicknesses):
+        return tdem.Simulation1DLayeredStitched(
+            survey=survey,
+            thicknesses=thicknesses,
+            sigmaMap=maps.ExpMap(nP=self.n_param(thicknesses)), 
+            parallel=self.parallel,
+            n_cpu=self.n_cpu,
+            n_layer=self.n_layer)
+    
+    def make_data(self, survey):
+        return data.Data(
+            survey,
+            dobs=self.data_array,
+            standard_deviation=self.uncert_array)
+    
+    def make_misfit_weights(self, thicknesses):
+        return 1./self.uncert_array
+
+    def make_misfit(self, thicknesses):
+        survey = self.make_survey()
+
+        dmis = data_misfit.L2DataMisfit(
+            simulation=self.make_simulation(survey, thicknesses),
+            data=self.make_data(survey))
+        dmis.W = self.make_misfit_weights(thicknesses)
+        return dmis
+    
+    alpha_s = 1e-10
+    alpha_r = 1.
+    alpha_z = 1.
+    def make_regularization(self, thicknesses):
+        if False:
+            assert False, "LCI is currently broken"
+            hz = np.r_[thicknesses, thicknesses[-1]]
+            reg = LaterallyConstrained(
+                get_2d_mesh(len(self.xyz.flightlines), hz),
+                mapping=maps.IdentityMap(nP=self.n_param(thicknesses)),
+                alpha_s = 0.01,
+                alpha_r = 1.,
+                alpha_z = 1.)
+            # reg.get_grad_horizontal(self.xyz.flightlines[["x", "y"]], hz, dim=2, use_cell_weights=True)
+            # ps, px, py = 0, 0, 0
+            # reg.norms = np.c_[ps, px, py, 0]
+            reg.mref = np.log(np.ones(self.n_param(thicknesses)) * 1/self.start_res)
+            # reg.mrefInSmooth = False
+            return reg
+        else:
+            coords = self.xyz.flightlines[[self.xyz.x_column, self.xyz.y_column]].astype(float).values
+            # FIXME: Triangulation fails if all coords are on a line, as in a typical synthetic case...
+            # Ideally, shouldn't this degrade gracefully to an LCI?
```

Comment:
Yes, it does. You can still invert all lines and just put constraints along the line by adjusting edge parameters. I'll provide an example code later.